File: src/sdk/cpp/libHanamiAiSdk/src/data_set.cpp
/**
 * @file        data_set.cpp
 *
 * @author      Tobias Anker <tobias.anker@kitsunemimi.moe>
 *
 * @copyright   Apache License Version 2.0
 *
 *      Copyright 2022 Tobias Anker
 *
 *      Licensed under the Apache License, Version 2.0 (the "License");
 *      you may not use this file except in compliance with the License.
 *      You may obtain a copy of the License at
 *
 *          http://www.apache.org/licenses/LICENSE-2.0
 *
 *      Unless required by applicable law or agreed to in writing, software
 *      distributed under the License is distributed on an "AS IS" BASIS,
 *      WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *      See the License for the specific language governing permissions and
 *      limitations under the License.
 */

#include <libHanamiAiSdk/data_set.h>
#include <libHanamiAiSdk/common/websocket_client.h>
#include <common/http_client.h>

#include <libKitsunemimiCrypto/common.h>
#include <libKitsunemimiJson/json_item.h>
#include <libKitsunemimiCommon/items/data_items.h>
#include <libKitsunemimiCommon/files/binary_file.h>

#include <../../../libraries/libKitsunemimiHanamiMessages/protobuffers/hanami_messages.proto3.pb.h>

namespace HanamiAI
{

/**
 * @brief get size of a local file
 *
 * @param filePath path the file to check
 *
 * @return size of file if successful, or -1 if failed
 */
long
getFileSize(const std::string &filePath)
{
    std::ifstream in(filePath, std::ifstream::ate | std::ifstream::binary);
    return in.tellg();
}

/**
 * @brief initialize a new csv-dataset in shiori
 *
 * @param result reference for response-message
 * @param dataSetName name for the new data-set
 * @param inputDataSize size of the file with the input-data
 * @param error reference for error-output
 *
 * @return true, if successful, else false
 */
bool
createCsvDataSet(std::string &result,
                 const std::string &dataSetName,
                 const uint64_t inputDataSize,
                 Kitsunemimi::ErrorContainer &error)
{
    // create request
    HanamiRequest* request = HanamiRequest::getInstance();
    const std::string path = "/control/v1/csv/data_set";
    const std::string vars = "";
    const std::string jsonBody = "{\"name\":\""    + dataSetName + "\""
                                 ",\"input_data_size\":" + std::to_string(inputDataSize) + "}";

    // send request
    if(request->sendPostRequest(result, path, vars, jsonBody, error) == false) {
        return false;
    }

    return true;
}

/**
 * @brief finalize data-set
 *
 * @param result reference for response-message
 * @param uuid uuid to identify the data-set
 * @param inputUuid uuid to identify the temporary file with the input-data on server-side
 * @param error reference for error-output
 *
 * @return true, if successful, else false
 */
bool
finalizeCsvDataSet(std::string &result,
                   const std::string &uuid,
                   const std::string &inputUuid,
                   Kitsunemimi::ErrorContainer &error)
{
    // create request
    HanamiRequest* request = HanamiRequest::getInstance();
    const std::string path = "/control/v1/csv/data_set";
    const std::string vars = "";
    const std::string jsonBody = "{\"uuid\":\""    + uuid + "\""
                                 ",\"uuid_input_file\":\"" + inputUuid + "\"}";

    // send request
    if(request->sendPutRequest(result, path, vars, jsonBody, error) == false) {
        return false;
    }

    return true;
}

/**
 * @brief initialize a new mnist-dataset in shiori
 *
 * @param result reference for response-message
 * @param dataSetName name for the new data-set
 * @param inputDataSize size of the file with the input-data
 * @param labelDataSize  size of the file with the label-data
 * @param error reference for error-output
 *
 * @return true, if successful, else false
 */
bool
createMnistDataSet(std::string &result,
                   const std::string &dataSetName,
                   const uint64_t inputDataSize,
                   const uint64_t labelDataSize,
                   Kitsunemimi::ErrorContainer &error)
{
    // create request
    HanamiRequest* request = HanamiRequest::getInstance();
    const std::string path = "/control/v1/mnist/data_set";
    const std::string vars = "";
    const std::string jsonBody = "{\"name\":\""    + dataSetName + "\""
                                 ",\"input_data_size\":" + std::to_string(inputDataSize) +
                                 ",\"label_data_size\":" + std::to_string(labelDataSize) + "}";

    // send request
    if(request->sendPostRequest(result, path, vars, jsonBody, error) == false) {
        return false;
    }

    return true;
}

/**
 * @brief finalize data-set
 *
 * @param result reference for response-message
 * @param uuid uuid to identify the data-set
 * @param inputUuid uuid to identify the temporary file with the input-data on server-side
 * @param labelUuid uuid to identify the temporary file with the label-data on server-side
 * @param error reference for error-output
 *
 * @return true, if successful, else false
 */
bool
finalizeMnistDataSet(std::string &result,
                     const std::string &uuid,
                     const std::string &inputUuid,
                     const std::string &labelUuid,
                     Kitsunemimi::ErrorContainer &error)
{
    // create request
    HanamiRequest* request = HanamiRequest::getInstance();
    const std::string path = "/control/v1/mnist/data_set";
    const std::string vars = "";
    const std::string jsonBody = "{\"uuid\":\""    + uuid + "\""
                                 ",\"uuid_input_file\":\"" + inputUuid + "\""
                                 ",\"uuid_label_file\":\"" + labelUuid + "\"}";

    // send request
    if(request->sendPutRequest(result, path, vars, jsonBody, error) == false) {
        return false;
    }

    return true;
}

/**
 * @brief send data to shiori
 *
 * @param session session over which the data should be send
 * @param datasetUuid uuid of the dataset where the file belongs to
 * @param fileUuid uuid of the file for identification in shiori
 * @param filePath path to file, which should be send
 * @param error reference for error-output
 *
 * @return true, if successful, else false
 */
bool
sendFile(WebsocketClient* client,
         const std::string &datasetUuid,
         const std::string &fileUuid,
         const std::string &filePath,
         Kitsunemimi::ErrorContainer &error)
{
    // get data, which should be send
    const uint64_t dataSize = getFileSize(filePath);
    Kitsunemimi::BinaryFile sourceFile(filePath);

    bool success = true;
    uint64_t pos = 0;

    // prepare buffer
    uint64_t segmentSize = 96 * 1024;

    uint8_t readBuffer[96*1024];
    uint8_t sendBuffer[128*1024];

    do
    {
        // check the size for the last segment
        segmentSize = 96 * 1024;
        if(dataSize - pos < segmentSize) {
            segmentSize = dataSize - pos;
        }

        FileUpload_Message message;
        message.set_fileuuid(fileUuid);
        message.set_datasetuuid(datasetUuid);
        message.set_type(UploadDataType::DATASET_TYPE);
        message.set_islast(false);
        if(pos + segmentSize >= dataSize) {
            message.set_islast(true);
        }

        // read segment of the local file
        if(sourceFile.readDataFromFile(&readBuffer[0], pos, segmentSize, error) == false)
        {
            success = false;
            error.addMeesage("Failed to read file '" + filePath + "'");
            break;
        }

        message.set_position(pos);
        message.set_data(readBuffer, segmentSize);

        const uint64_t msgSize = message.ByteSizeLong();
        if(message.SerializeToArray(sendBuffer, msgSize) == false)
        {
            error.addMeesage("Failed to serialize train-message");
            return false;
        }

        // send segment
        if(client->sendMessage(sendBuffer, msgSize, error) == false)
        {
            LOG_ERROR(error);
            success = false;
            break;
        }

        pos += segmentSize;
    }
    while(pos < dataSize);

    return success;
}

/**
 * @brief wait until the upload of all tempfiles are complete
 *
 * @param uuid uuid of the dataset
 * @param error reference for error-output
 *
 * @return true, if successful, else false
 */
bool
waitUntilFullyUploaded(const std::string &uuid,
                       Kitsunemimi::ErrorContainer &error)
{
    // TODO: add timeout-timer
    bool completeUploaded = false;
    while(completeUploaded == false)
    {
        sleep(1);

        std::string progressStr = "";
        if(getDatasetProgress(progressStr, uuid, error) == false)
        {
            LOG_ERROR(error);
            return false;
        }

        Kitsunemimi::JsonItem progress;
        if(progress.parse(progressStr, error) == false)
        {
            LOG_ERROR(error);
            return false;
        }

        completeUploaded = progress.get("complete").getBool();
    }

    return true;
}

/**
 * @brief upload new csv-data-set to shiori
 *
 * @param result reference for response-message
 * @param dataSetName name for the new data-set
 * @param inputFilePath path to file with the inputs
 * @param error reference for error-output
 *
 * @return true, if successful, else false
 */
bool
uploadCsvData(std::string &result,
              const std::string &dataSetName,
              const std::string &inputFilePath,
              Kitsunemimi::ErrorContainer &error)
{
    // init new mnist-data-set
    if(createCsvDataSet(result,
                        dataSetName,
                        getFileSize(inputFilePath),
                        error) == false)
    {
        return false;
    }

    // parse output to get the uuid
    Kitsunemimi::JsonItem jsonItem;
    if(jsonItem.parse(result, error) == false)
    {
        LOG_ERROR(error);
        return false;
    }

    // get ids from inital reponse to identify the file-transfer
    const std::string uuid = jsonItem.get("uuid").getString();
    const std::string inputUuid = jsonItem.get("uuid_input_file").getString();

    // init websocket to shiori
    WebsocketClient wsClient;
    std::string websocketUuid = "";
    const bool ret = wsClient.initClient(websocketUuid,
                                         HanamiRequest::getInstance()->getToken(),
                                         "shiori",
                                         HanamiRequest::getInstance()->getHost(),
                                         HanamiRequest::getInstance()->getPort(),
                                         "",
                                         error);
    if(ret == false)
    {
        error.addMeesage("Failed to init websocket to shiori");
        LOG_ERROR(error);
        return false;
    }

    // send file
    if(sendFile(&wsClient, uuid, inputUuid, inputFilePath, error) == false)
    {
        LOG_ERROR(error);
        return false;
    }

    // wait until all data-transfers to shiori are completed
    if(waitUntilFullyUploaded(uuid, error) == false)
    {
        LOG_ERROR(error);
        return false;
    }

    if(finalizeCsvDataSet(result, uuid, inputUuid, error) == false)
    {
        LOG_ERROR(error);
        return false;
    }

    return true;
}

/**
 * @brief upload new mnist-data-set to shiori
 *
 * @param result reference for response-message
 * @param dataSetName name for the new data-set
 * @param inputFilePath path to file with the inputs
 * @param labelFilePath path to file with the labels
 * @param error reference for error-output
 *
 * @return true, if successful, else false
 */
bool
uploadMnistData(std::string &result,
                const std::string &dataSetName,
                const std::string &inputFilePath,
                const std::string &labelFilePath,
                Kitsunemimi::ErrorContainer &error)
{
    // init new mnist-data-set
    if(createMnistDataSet(result,
                          dataSetName,
                          getFileSize(inputFilePath),
                          getFileSize(labelFilePath),
                          error) == false)
    {
        return false;
    }

    // parse output to get the uuid
    Kitsunemimi::JsonItem jsonItem;
    if(jsonItem.parse(result, error) == false)
    {
        LOG_ERROR(error);
        return false;
    }

    // get ids from inital reponse to identify the file-transfer
    const std::string uuid = jsonItem.get("uuid").getString();
    const std::string inputUuid = jsonItem.get("uuid_input_file").getString();
    const std::string labelUuid = jsonItem.get("uuid_label_file").getString();

    // init websocket to shiori
    WebsocketClient wsClient;
    std::string websocketUuid = "";
    const bool ret = wsClient.initClient(websocketUuid,
                                         HanamiRequest::getInstance()->getToken(),
                                         "shiori",
                                         HanamiRequest::getInstance()->getHost(),
                                         HanamiRequest::getInstance()->getPort(),
                                         "",
                                         error);
    if(ret == false)
    {
        error.addMeesage("Failed to init websocket to shiori");
        LOG_ERROR(error);
        return false;
    }

    // send file with inputs
    if(sendFile(&wsClient, uuid, inputUuid, inputFilePath, error) == false)
    {
        error.addMeesage("Failed to send file with input-values");
        LOG_ERROR(error);
        return false;
    }

    // send file with labels
    if(sendFile(&wsClient, uuid, labelUuid, labelFilePath, error) == false)
    {
        error.addMeesage("Failed to send file with labes");
        LOG_ERROR(error);
        return false;
    }

    // wait until all data-transfers to shiori are completed
    if(waitUntilFullyUploaded(uuid, error) == false)
    {
        error.addMeesage("Failed to wait for fully uploaded files");
        LOG_ERROR(error);
        return false;
    }

    if(finalizeMnistDataSet(result, uuid, inputUuid, labelUuid, error) == false)
    {
        error.addMeesage("Failed to finalize MNIST-dataset");
        LOG_ERROR(error);
        return false;
    }

    return true;
}

/**
 * @brief check values against a data-set to get correctness
 *
 * @param result reference for response-message
 * @param dataUuid uuid of the data-set to compare to
 * @param resultUuid uuid of the result-set to compare
 * @param error reference for error-output
 *
 * @return true, if successful, else false
 */
bool
checkDataset(std::string &result,
             const std::string &dataUuid,
             const std::string &resultUuid,
             Kitsunemimi::ErrorContainer &error)
{
    // create request
    HanamiRequest* request = HanamiRequest::getInstance();
    const std::string path = "/control/v1/data_set/check";
    const std::string vars = "";
    const std::string jsonBody = "{\"data_set_uuid\":\"" + dataUuid + "\""
                                 ",\"result_uuid\":\"" + resultUuid + "\"}";

    // send request
    if(request->sendPostRequest(result, path, vars, jsonBody, error) == false) {
        return false;
    }

    return true;
}

/**
 * @brief get metadata of a specific data-set
 *
 * @param result reference for response-message
 * @param dataUuid uuid of the requested data-set
 * @param error reference for error-output
 *
 * @return true, if successful, else false
 */
bool
getDataset(std::string &result,
           const std::string &dataUuid,
           Kitsunemimi::ErrorContainer &error)
{
    // create request
    HanamiRequest* request = HanamiRequest::getInstance();
    const std::string path = "/control/v1/data_set";
    const std::string vars = "uuid=" + dataUuid;

    // send request
    if(request->sendGetRequest(result, path, vars, error) == false)
    {
        error.addMeesage("Failed to get dataset with UUID '" + dataUuid + "'");
        LOG_ERROR(error);
        return false;
    }

    return true;
}

/**
 * @brief list all data-sets of the user, which are available on shiori
 *
 * @param result reference for response-message
 * @param error reference for error-output
 *
 * @return true, if successful, else false
 */
bool
listDatasets(std::string &result,
             Kitsunemimi::ErrorContainer &error)
{
    // create request
    HanamiRequest* request = HanamiRequest::getInstance();
    const std::string path = "/control/v1/data_set/all";

    // send request
    if(request->sendGetRequest(result, path, "", error) == false)
    {
        error.addMeesage("Failed to list datasets");
        LOG_ERROR(error);
        return false;
    }

    return true;
}

/**
 * @brief delete a data-set from shiori
 *
 * @param result reference for response-message
 * @param dataUuid uuid of the data-set to delete
 * @param error reference for error-output
 *
 * @return true, if successful, else false
 */
bool
deleteDataset(std::string &result,
              const std::string &dataUuid,
              Kitsunemimi::ErrorContainer &error)
{
    // create request
    HanamiRequest* request = HanamiRequest::getInstance();
    const std::string path = "/control/v1/data_set";
    const std::string vars = "uuid=" + dataUuid;

    // send request
    if(request->sendDeleteRequest(result, path, vars, error) == false)
    {
        error.addMeesage("Failed to delete dataset with UUID '" + dataUuid + "'");
        LOG_ERROR(error);
        return false;
    }

    return true;
}

/**
 * @brief check progress of file-upload
 *
 * @param result reference for response-message
 * @param dataUuid uuid of the data-set to get
 * @param error reference for error-output
 *
 * @return true, if successful, else false
 */
bool
getDatasetProgress(std::string &result,
                   const std::string &dataUuid,
                   Kitsunemimi::ErrorContainer &error)
{
    // create request
    HanamiRequest* request = HanamiRequest::getInstance();
    const std::string path = "/control/v1/data_set/progress";
    const std::string vars = "uuid=" + dataUuid;

    // send request
    if(request->sendGetRequest(result, path, vars, error) == false)
    {
        error.addMeesage("Failed to check upload-state of dataset with UUID '" + dataUuid + "'");
        LOG_ERROR(error);
        return false;
    }

    return true;
}

} // namespace HanamiAI

File: src/sdk/python/hanami_sdk/hanami_sdk/dataset.py
# Copyright 2022 Tobias Anker
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from . import hanami_request
from . import hanami_exceptions
from .hanami_messages import proto3_pb2
import json
import math
import time
import asyncio
import ssl
import websockets


def list_datasets(token: str,
                  address: str,
                  verify_connection: bool = True) -> str:
    path = "/control/v1/dataset/all"
    return hanami_request.send_get_request(token,
                                           address,
                                           path,
                                           "",
                                           verify=verify_connection)


def get_dataset(token: str,
                address: str,
                dataset_uuid: str,
                verify_connection: bool = True) -> str:
    path = "/control/v1/dataset"
    values = f'uuid={dataset_uuid}'
    return hanami_request.send_get_request(token,
                                           address,
                                           path,
                                           values,
                                           verify=verify_connection)


def delete_dataset(token: str,
                   address: str,
                   checkpoint_uuid: str,
                   verify_connection: bool = True) -> str:
    path = "/control/v1/dataset"
    values = f'uuid={checkpoint_uuid}'
    return hanami_request.send_delete_request(token,
                                              address,
                                              path, 
                                              values, 
                                              verify=verify_connection)


def check_mnist_dataset(token: str,
                        address: str,
                        dataset_uuid: str,
                        result_dataset_uuid: str,
                        verify_connection: bool = True) -> str:
    path = "/control/v1/dataset/check"
    values = f'dataset_uuid={dataset_uuid}&result_uuid={result_dataset_uuid}'
    return hanami_request.send_get_request(token,
                                           address,
                                           path, 
                                           values, 
                                           verify=verify_connection)


def wait_until_upload_complete(token: str,
                               address: str,
                               uuid: str,
                               verify_connection: bool = True) -> bool:
    while True:
        path = "/control/v1/dataset/progress"
        values = f'uuid={uuid}'

        result = hanami_request.send_get_request(token,
                                                 address,
                                                 path,
                                                 values,
                                                 verify=verify_connection)

        result_json = json.loads(result)
        if result_json["complete"]:
            return True

        time.sleep(0.5)

    return True


async def send_data(token: str,
                    address: str,
                    dataset_uuid: str,
                    file_uuid: str,
                    data,
                    verify_connection: bool = True) -> bool:
    # create initial request for the websocket-connection
    initial_ws_msg = {
        "token": token,
        "target": "file_upload",
        "uuid": file_uuid,
    }
    body_str = json.dumps(initial_ws_msg)

    ssl_context = None
    ws_begin = "ws"
    if address.startswith("https"):
        ws_begin = "wss"

        # Disable SSL verification
        if not verify_connection:
            ssl_context = ssl.SSLContext()
            ssl_context.verify_mode = ssl.CERT_NONE

    base_address = address.split('/')[2]

    async with websockets.connect(ws_begin + "://" + base_address, ssl=ssl_context) as websocket:
        await websocket.send(body_str)
        message = await websocket.recv()
        result_json = json.loads(message)

        if result_json["success"] is False:
            return False

        total_size = len(data)
        chunk_size = 128 * 1024
        num_chunks = math.ceil(total_size / chunk_size)

        for i in range(num_chunks):
            start = i * chunk_size
            end = min((i + 1) * chunk_size, total_size)
            chunk_data = data[start:end]

            file_upload_msg = proto3_pb2.FileUpload_Message()
            file_upload_msg.position = start
            file_upload_msg.data = chunk_data
            serialized_msg = file_upload_msg.SerializeToString()

            await websocket.send(serialized_msg)
            message = await websocket.recv()

    return True


def upload_mnist_files(token: str,
                       address: str,
                       name: str,
                       input_file_path: str,
                       label_file_path: str,
                       verify_connection: bool = True) -> str:
    # read files
    with open(input_file_path, 'rb') as i_f:
        input_file_data = i_f.read()

    with open(label_file_path, 'rb') as l_f:
        label_file_data = l_f.read()

    # initialize
    path = "/control/v1/dataset/upload/mnist"
    json_body = {
        "name": name,
        "input_data_size": len(input_file_data),
        "label_data_size": len(label_file_data),
    }
    body_str = json.dumps(json_body)
    result = hanami_request.send_post_request(
        token, address, path, body_str, verify=verify_connection)

    # process init-result
    result_json = json.loads(result)
    uuid = result_json["uuid"]
    input_file_uuid = result_json["uuid_input_file"]
    label_file_uuid = result_json["uuid_label_file"]

    # send data
    if not asyncio.run(send_data(token,
                                 address,
                                 uuid,
                                 input_file_uuid,
                                 input_file_data,
                                 verify_connection=verify_connection)):
        raise hanami_exceptions.InternalServerErrorException()

    if not asyncio.run(send_data(token,
                                 address,
                                 uuid,
                                 label_file_uuid,
                                 label_file_data,
                                 verify_connection=verify_connection)):
        raise hanami_exceptions.InternalServerErrorException()

    if not wait_until_upload_complete(token,
                                      address,
                                      uuid,
                                      verify_connection=verify_connection):
        raise hanami_exceptions.InternalServerErrorException()

    # finalize
    path = "/control/v1/dataset/upload/mnist"
    json_body = {
        "uuid": uuid,
        "uuid_input_file": input_file_uuid,
        "uuid_label_file": label_file_uuid,
    }
    body_str = json.dumps(json_body)
    result = hanami_request.send_put_request(
        token, address, path, body_str, verify=verify_connection)
    result_json = json.loads(result)

    return uuid


def upload_csv_files(token: str,
                     address: str,
                     name: str,
                     input_file_path: str,
                     verify_connection: bool = True) -> str:
    # read files
    with open(input_file_path, 'rb') as i_f:
        input_file_data = i_f.read()

    # initialize
    path = "/control/v1/dataset/upload/csv"
    json_body = {
        "name": name,
        "input_data_size": len(input_file_data),
    }
    body_str = json.dumps(json_body)
    result = hanami_request.send_post_request(token,
                                              address,
                                              path,
                                              body_str,
                                              verify=verify_connection)

    # process init-result
    result_json = json.loads(result)
    uuid = result_json["uuid"]
    input_file_uuid = result_json["uuid_input_file"]

    # send data
    if not asyncio.run(send_data(token,
                                 address,
                                 uuid,
                                 input_file_uuid,
                                 input_file_data,
                                 verify_connection=verify_connection)):
        raise hanami_exceptions.InternalServerErrorException()

    if not wait_until_upload_complete(token,
                                      address,
                                      uuid,
                                      verify_connection=verify_connection):
        raise hanami_exceptions.InternalServerErrorException()

    # finalize
    path = "/control/v1/dataset/upload/csv"
    json_body = {
        "uuid": uuid,
        "uuid_input_file": input_file_uuid,
    }
    body_str = json.dumps(json_body)
    result = hanami_request.send_put_request(
        token, address, path, body_str, verify=verify_connection)
    result_json = json.loads(result)

    return uuid
